### PR TITLE
Add filters for Buf Connect RPC protocol.

### LINKED ISSUE #22679: Add support for Connect RPC protocol
*Title*: *Add support for the Connect RPC protocol, on par with gRPC and gRPC-Web*

*Description*:
Add two features to support the [Connect RPC protocol](https://connect.build/docs/protocol):

1. Export metrics for proxied Connect traffic, as Envoy does today for gRPC.
2. Add a filter to translate from the Connect protocol to gRPC, as Envoy does today to expose gRPC backends to gRPC-Web clients.

The Connect protocol and the associated RPC libraries are similar to gRPC: they use HTTP as a transport, support unary and streaming RPCs, and are encoding agnostic but usually paired with Protobuf. Unlike gRPC, the Connect protocol works with HTTP/1.1,<sup>&dagger;</sup> doesn't use HTTP trailers, and keeps the wire representation of unary RPCs [dead simple](https://connect.build/docs/protocol#unary-examples). Because it demands very little of the underlying HTTP transport, Connect is easy to adopt: clients can be as simple as a cURL one-liner, and popular frameworks like Rails, Flask, and Laravel can serve Connect traffic.

Envoy's excellent metrics are, by themselves, a significant incentive for organizations to adopt gRPC. It'd be wonderful to bring the same consistent observability to Connect traffic.

Of course, many organizations already have large gRPC deployments. A translating filter would let web browsers call these backends using a protocol that's more developer-friendly than gRPC-Web. We designed Connect with translating proxies in mind, so converting to and from gRPC is simple, reasonably efficient, and semantically lossless.

Thanks for considering this enhancement!

*Relevant Links*:
- The [Connect website](https://connect.build), especially the [project overview](https://connect.build/docs/introduction)
- The [Connect protocol specification](https://connect.build/docs/protocol)
- The [Go implementation](https://github.com/bufbuild/connect-go)
- The [TypeScript implementation](https://github.com/bufbuild/connect-web) (for browsers)
- Implementations for Node.js, Kotlin, and Swift are in progress.

<hr>
<sup>&dagger;</sup> Naturally, bidirectional streaming still requires HTTP/2. Unary, server streaming, and client streaming RPCs work over HTTP/1.1 or HTTP/2.

### PR DESCRIPTION
**Pull request by jchadwick-buf**
Signed-off-by: John Chadwick <jchadwick@buf.build>

Implements two Connect filters:
- `connect_stats`: Similar to `grpc_stats`. Counts messages and success/failure as stats.
- `connect_grpc_bridge`: Provides a bridge from Connect clients to gRPC servers.

The [Buf Connect](https://connect.build/) protocol provides similar features to gRPC, but supports simple HTTP/1.1 requests for unary procedure calls and does not require trailers for streaming procedure calls.

The stats filter is similar to `grpc_stats`, as mentioned prior. It could potentially be rolled into `grpc_stats`, although the Connect protocol does differ from gRPC in a fair few ways, so I do anticipate the resulting code would be a little hard to follow. The bridge filter is not exactly analogous to any existing gRPC filters. For streaming Buf Connect requests, it is somewhat similar to the `grpc_web` filter, but for unary Buf Connect requests, it's more similar to `grpc_http1_bridge`.

The bridge filter is mainly useful for customers who are interested in transitioning to using Connect incrementally, whereas the stats filter is useful for anyone who wants to ensure that their stats covers usage of the Connect protocol, either independently or in addition to gRPC.

Closes #22679.

### CODE REVIEW

**Comment by akshayjshah** [comment]
@mattklein123 appreciate you taking the time to chat with us today and maybe sponsor this contribution. @jchadwick-buf is going to highlight areas of the diff which could use a maintainer's eye and point out the few areas where we still have a little work to do.

File: source/extensions/filters/http/connect_grpc_bridge/end_stream_response.cc
Comment (jchadwick-buf) [inline source/extensions/filters/http/connect_grpc_bridge/end_stream_response.cc:76]: I think it's fair to say that this is not an ideal way to deal with serializing JSON. I am open to rewriting this using a different approach if there is a preferable one that would work for our use case.

File: source/extensions/filters/http/connect_stats/connect_stats_filter.cc
Comment (jchadwick-buf) [inline source/extensions/filters/http/connect_stats/connect_stats_filter.cc:28]: The duplication here may hint that it is better to simply merge this filter with `grpc_stats`. I'm open to that perspective even if the bridge filter remains independent. Of course, if there's another good way to reduce this duplication, I'm all-ears.

Comment (mattklein123) [reply source/extensions/filters/http/connect_stats/connect_stats_filter.cc:28]: I would vote to merge it, yes. I think there is so much overlap in config, etc. that it doesn't make sense to have a new filter for this part.

File: source/extensions/filters/http/connect_stats/response_frame_counter.cc
Comment (mattklein123) [reply source/extensions/filters/http/connect_stats/response_frame_counter.cc:48]: It would be better to just make a variant of `loadFromString` that returns a status/result.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
At a high level this makes sense. I left some comments and I'm happy to sponsor. Thanks.

/wait

File: source/extensions/filters/http/connect_grpc_bridge/end_stream_response.cc
Comment (jchadwick-buf) [reply source/extensions/filters/http/connect_grpc_bridge/end_stream_response.cc]: After some wrangling, Proto WKT works for both this use case as well as the JSON decoding we do in stats, so it has been replaced. Now both use Proto WKT. Resolving.

File: source/extensions/filters/http/connect_stats/connect_stats_filter.cc
Comment (jchadwick-buf) [reply source/extensions/filters/http/connect_stats/connect_stats_filter.cc]: Merged. Thanks.

File: source/extensions/filters/http/connect_stats/response_frame_counter.cc
Comment (jchadwick-buf) [reply source/extensions/filters/http/connect_stats/response_frame_counter.cc]: Resolved by replacing with Protobuf WKT usage, which doesn't have this issue. Resolving thread.

File: source/extensions/filters/http/connect_grpc_bridge/config.cc
Comment (jchadwick-buf) [reply source/extensions/filters/http/connect_grpc_bridge/config.cc]: I decided to put in the effort to merge them. It was definitely a bit tricky but I think the resulting filter is not too hard to read after all, so it was probably worth it. I am considering this thread resolved.

**Comment by jchadwick-buf** [comment]
I don't know why it says I added and then removed reviewers: I didn't, I only hit "Ready for Review." In any case, I think this is ready for review now. I've tested it in many scenarios and I've tried to cover the code with an adequate amount of tests. If desired, I can split this into more PRs, merge/rebase commits, etc. I tried to avoid too much rebasing for the moment so that the progres/changes were not obscured.

File: api/envoy/extensions/filters/http/grpc_stats/v3/config.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/http/grpc_stats/v3/config.proto:76]: I might be missing something here, but the filter is for GrpcStats, and Buf-Connect is a different protocol, so shouldn't there be a BufConnectStats filter (that has the relevant config)?

Comment (jchadwick-buf) [reply api/envoy/extensions/filters/http/grpc_stats/v3/config.proto:76]: What we wound up doing was merging Buf Connect support into the gRPC stats protocol due to the overlap between the protocols. (Connect is semantically compatible with gRPC, so it's somewhat akin to say, gRPC-web.) I'm open to trying to make this make more sense. Context: https://github.com/envoyproxy/envoy/pull/24836#discussion_r1065012094

Comment (mattklein123) [reply api/envoy/extensions/filters/http/grpc_stats/v3/config.proto:76]: Given the similarity of the protocols I think it's reasonable to have it be part of this filter. We can potentially make this more clear in the docs. I will take a look.

**Comment by mattklein123** [comment]
Sorry for the review delay. I will get back to this within the next week.

File: api/envoy/extensions/filters/http/grpc_stats/v3/config.proto
Comment (mattklein123) [reply api/envoy/extensions/filters/http/grpc_stats/v3/config.proto:76]: Is there any reason to actually have this be configurable? It's turned on by content-type, right? Can we just turn it on automatically and then just document that it is supported somewhere in the top level filter documentation?

File: docs/root/configuration/http/http_filters/connect_grpc_bridge_filter.rst
Comment (mattklein123) [inline docs/root/configuration/http/http_filters/connect_grpc_bridge_filter.rst:1]: Can you also update relevant docs here: https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_protocols/grpc

And ref link back to this filter where appropriate.

File: test/extensions/filters/http/connect_grpc_bridge/connect_grpc_bridge_integration_test.cc
Comment (mattklein123) [inline test/extensions/filters/http/connect_grpc_bridge/connect_grpc_bridge_integration_test.cc:8]: Can you have an integration test for the actual end to end flow for both unary and streaming?

In general can you also check the code coverage report and make sure you are reasonably covering all of the code?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks at a high level this LGTM. Given the alpha status I think it's fine to ship and iterate with some real customers.

Note that if we want to call this eventually robust to untrusted data we should have some limited fuzzing of all of the relevant parsing code. This can be done later.

Thank you!

/wait

File: api/envoy/extensions/filters/http/grpc_stats/v3/config.proto
Comment (jchadwick-buf) [reply api/envoy/extensions/filters/http/grpc_stats/v3/config.proto]: Done.

File: docs/root/configuration/http/http_filters/connect_grpc_bridge_filter.rst
Comment (jchadwick-buf) [reply docs/root/configuration/http/http_filters/connect_grpc_bridge_filter.rst:1]: I added the bridge to the list, is that what you mean? Or should it have a section just for connect?

File: test/extensions/filters/http/connect_grpc_bridge/connect_grpc_bridge_integration_test.cc
Comment (jchadwick-buf) [reply test/extensions/filters/http/connect_grpc_bridge/connect_grpc_bridge_integration_test.cc:12]: Done.

I did try to ensure the important cases are covered, and we have >96% on everything. We can aim for 100% if you would prefer though. I think the main tricky case is trying to find a way to get WKT JSON encoding to fail.

**Review by mattklein123 — APPROVED** [review]
Great, thanks!

**Comment by jchadwick-buf** [comment]
Thanks! Github says workflows need approval still. Did I need to do something there?

**Pull request merged**
